### PR TITLE
[stable/redmine] Add JSON Schema & standardize ingress configuration

### PR DESCRIPTION
**Pull request by juan131**
Signed-off-by: juan131 <juan@bitnami.com>

#### Is this a new chart

No

**Description of the change**

This PR includes a basic JSON schema for the Redmine chart. It doesn't cover all the possible values of the chart but just the ones that, in my opinion, are more likely to be modified.

This is valid for Helm 3 to validate some values and, at the same time, for Kubeapps to render a simple form so the chart containing it is easier to configure and deploy. See https://github.com/kubeapps/kubeapps/blob/master/docs/developer/basic-form-support.md for more information.

This is an example of the file rendered by Kubeapps:

Screenshot 2019-12-02 at 13 50 51

This PR also standardizes the way we configure Ingress.

**Checklist**

- [x] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [x] Title of the PR starts with chart name (e.g. `[stable/chart]`)

### CODE REVIEW

**Comment by juan131** [comment]
/hold

Blocking this PR since the syntax below in the `hidden` section is not supported yet (see https://github.com/kubeapps/kubeapps/blob/master/docs/developer/basic-form-support.md#subsections):

```json
    "externalDatabase": {
      "type": "object",
      "title": "External Database Details",
      "description": "If MariaDB is disabled. Use this section to specify the external database details",
      "form": true,
      "hidden": "mariadb.enabled and postgresql.enabled",
```

File: stable/redmine/values.schema.json
Comment (andresmgot) [inline stable/redmine/values.schema.json:88]: maybe this is a problem of the chart but why is there `databaseType.mariadb` and `mariadb.enabled`? Isn't that for the same thing?

Also, in case you want to keep `databaseType`, you may want to make it a string with two allowed values: `mariadb|postgresql|external`. You can force that with the verification :)

File: stable/redmine/values.schema.json
Comment (andresmgot) [inline stable/redmine/values.schema.json]: My recommendation is to remove this parameter so we unblock this PR, making a note to add it back once the linked issue is addressed.

**Review by andresmgot — COMMENTED** [review]
Thanks @juan131 . See my comments and let me know what you think.

File: stable/redmine/values.schema.json
Comment (juan131) [reply stable/redmine/values.schema.json:88]: > maybe this is a problem of the chart but why is there databaseType.mariadb and mariadb.enabled? Isn't that for the same thing?

We use `databaseType` to decide whether to use MariaDB or PostgreSQL as database. On the other hand, we use `mariadb.enabled` to decide whether we should use the MariaDB chart to create a database in the cluster or not.

> Also, in case you want to keep databaseType, you may want to make it a string with two allowed values: mariadb|postgresql|external. You can force that with the verification :)

That's a good suggestion. I'll apply it

Comment (andresmgot) [reply stable/redmine/values.schema.json:88]: > > maybe this is a problem of the chart but why is there databaseType.mariadb and mariadb.enabled? Isn't that for the same thing?
> 
> We use `databaseType` to decide whether to use MariaDB or PostgreSQL as database. On the other hand, we use `mariadb.enabled` to decide whether we should use the MariaDB chart to create a database in the cluster or not.
> 

What would happen if you set `databaseType.mariadb` and `mariadb.enabled=false` at the same time? Isn't that an inconsistent status? What I mean, is there a valid use case in which those two parameters may not match?

**Comment by juan131** [comment]
No backwards incompatibility introduce since the commands below work:

```
$ helm install stable/redmine --name redmine \
    --set redminePassword=myredminepass \
    --set mariadb.rootUser.password=myredminedbpass \
    --set databaseType.mariadb=true \
    --version 13.6.0
$ helm upgrade redmine stable/redmine \
    --set redminePassword=myredminepass \
    --set mariadb.rootUser.password=myredminedbpass \
    --set databaseType="mariadb" \
    --version 13.7.0
```

File: stable/redmine/values.schema.json
Comment (juan131) [reply stable/redmine/values.schema.json]: Check the issue now @andresmgot 
Using "databaseType" as a string gives us even less possibilities for the JSON schema

File: stable/redmine/values.schema.json
Comment (andresmgot) [inline stable/redmine/values.schema.json:93]: you can force the strings using the property `enum`:

```
"enum": ["mariadb", "postgresql"]
```

Also, why don't you add "external"? That way we unblock this schema.

Apart from that, you didn't reply to my question: What would happen if you set databaseType.mariadb and mariadb.enabled=false at the same time? Isn't that an inconsistent status? What I mean, is there a valid use case in which those two parameters may not match?

File: stable/redmine/values.schema.json
Comment (andresmgot) [inline stable/redmine/values.schema.json]: here it should be just `{"condition": "mariadb", "value": "databaseType"}`

File: stable/redmine/values.schema.json
Comment (andresmgot) [inline stable/redmine/values.schema.json]: same here about codition/value

File: stable/redmine/Chart.yaml
Comment (andresmgot) [inline stable/redmine/Chart.yaml]: I would bump the major version here.

Comment (juan131) [reply stable/redmine/Chart.yaml]: I will support both ways to setup Ingress

File: stable/redmine/values.schema.json
Comment (juan131) [reply stable/redmine/values.schema.json:93]: If you set that, you'll face an error (check NOTES.txt)

**Comment by sameersbn** [comment]
@juan131  needs rebase

**Comment by juan131** [comment]
@sameersbn rebase done

File: stable/redmine/values.yaml
Comment (sameersbn) [inline stable/redmine/values.yaml]: Do you think it makes sense to have a validation for this in the NOTES.txt?

File: stable/redmine/values.yaml
Comment (sameersbn) [inline stable/redmine/values.yaml:81]: Does this change warrant a major version bump? At any rate we should document this change in the README

**Review by sameersbn — APPROVED** [review]
have some comments, overall lgtm 
/ok-to-test

File: stable/redmine/values.yaml
Comment (juan131) [reply stable/redmine/values.yaml:81]: If we are strict, yes, it requires a major version bump. We can:
- a) Bum the major version (removing the support to the old way to configure ingress)
- b) Document the change in the README.md
- c) Both of them

Comment (juan131) [reply stable/redmine/values.yaml:81]: What do you prefer?

Comment (sameersbn) [reply stable/redmine/values.yaml:81]: I would prefer both of them

**Review by andresmgot — APPROVED** [review]
My +1,

**Review by sameersbn — APPROVED** [review]

**Pull request merged**
